### PR TITLE
Fix/improve error locations

### PR DESCRIPTION
**Pull request by christopherswenson**
This PR attaches correct (or at least, more correct) locations to errors. Firstly, it fixes an issue from https://github.com/looker-open-source/malloy/pull/364 where semantic errors appeared on the line before they were supposed to. It also adjusts the location of syntax errors to only include the offending token, and not the offending token plus the rest of that line.

### CODE REVIEW

File: packages/malloy/src/lang/parse-malloy.ts [hunk 1]
```diff
@@ -61,13 +62,16 @@ class ParseErrorHandler implements ANTLRErrorListener<Token> {
...
+          start: { line: 0, character: 0 },
+          end: { line: 0, character: 0 },
+        };
     const error: LogMessage = {
-      sourceURL: this.sourceURL,
       message: msg,
-      begin: {
-        line: line,
-        char: charPositionInLine,
-      },
```
Comment (christopherswenson) [inline packages/malloy/src/lang/parse-malloy.ts:70]: There was previously no `end` specified on syntax errors, so the IDE (language server diagnostic code, specifically) was filling in an `end` of the end of the line.

**Comment by mtoy-googly-moogly** [comment]
ok, i broke this fix with my PR -- fixing the broken fix, i should have tested before i pushed.

**Comment by mtoy-googly-moogly** [comment]
Fixing it by adding tests with `markSource` 🥳  because @christopherswenson is amazing!

**Comment by mtoy-googly-moogly** [comment]
@christopherswenson this makes ALL the location reference code in lang, error messages AND the external reference locations, use the DocumentLocation datatype.

Added a couple of tests to see that error messages highlight properly.

Ready for review.

**Comment by christopherswenson** [comment]
@mtoy-googly-moogly Looks good to me 👍 Thanks for the help here!

**Pull request merged**